RIOT-OS/RIOT
# periph/adc: support for ADC extension API

### Contribution description

This is an implementation of the extension API as proposed in #9582 for ADC extensions. It corresponds to the implementation of the GPIO extension API in #9860 and #9958. The PR contains the following contributions:

- Low level function prototypes `adc_*_ll` in `drivers/include/periph.h`.
- Changes of `adc_*` functions to redirect a function call either to the low level functions `adc_*_ll` or to the `adc_ext_*` function provided by an ADC extension device driver.
- New inline function `adc_channels` which returns the number of channels of a device, in case of MCU channels simply `ADC_NUMOF`. This new function makes it possible to get the information from a certain device how many channels it has and allows to iterate over the all channels of all devices.
- Changes of low level function `adc_*_ll` for all CPUs that support ADC channels.
- Boards that base on SAM0 (SAMD21, SAML21) MCU the ADC channel configuration `adc_channels` was renamed to `adc_config` to solve conflicts with the new function `adc_channels`.

The ADC extension API does not require feature `periph_adc` which makes it possible to extend boards that do not provide ADC capabilities by the MCU by external ADC modules.

### Testing procedure

A test case is provided that implements a soft-driver for the ADC extension interface. The test case uses this driver to confirm that interception and redirection of the API call are working properly. This has been tested and is working properly on esp8266-esp-12x, esp32-wroom-32 and arduino-mega2560.

To test only the ADC extension API use 
```
make  flash test -C tests/extend_adc BOARD=...
```
To test the ADC extension API together with internal ADC channels use
```
USEMODULE=periph_adc make  flash test -C tests/extend_adc BOARD=...
```
In that case feature `periph_adc` is required of course.

### Issues/PRs references

Implements #9582 
Corresponds to #9860 and #9958 

The changes in the `makefiles/pseudomodules.mk` and  in `drivers/Makefile.include` made to get it working are the same as in #9958. These changes might be removed once #9958 is merged.
